book/src/dev/rfcs/drafts/xxxx-async-rust.md

```diff
@@ -0,0 +1,438 @@
+
+Ideally, buffer slots or mutexes should be acquired one at a time, and held for as short a time
+as possible.
+
+If that's not possible, acquire slots and mutexs in the same order across cooperating tasks.
+If tasks acquire the same slots and mutexes in different orders, they can deadlock, each holding
+a lock that the other needs.
+
+Note: do not call `poll_ready` on multiple tasks, then match against the results. Use the `ready!`
+macro instead.
+
+If a task is waiting for service readiness, and that service depends on other futures to become ready,
+make sure those futures are scheduled in separate tasks using `tokio::spawn`.
+
+## Buffer and Batch
+[buffer-and-batch]: #buffer-and-batch
+
+The constraints imposed by the `tower::Buffer` and `tower::Batch` implementations are:
+1. `poll_ready` must be called **at least once** for each `call`
+2. Once we've reserved a buffer slot, we always get `Poll::Ready` from a buffer, regardless of the
+   current readiness of the buffer or its underlying service
+3. The `Buffer`/`Batch` capacity limits the number of concurrently waiting tasks. Once this limit
+   is reached, further tasks will block, awaiting a free reservation.
+4. Some tasks can depend on other tasks before they resolve. (For example: block validation.)
+   If there are task dependencies, **the `Buffer`/`Batch` capacity must be larger than the
+   maximum number of concurrently waiting tasks**, or Zebra could deadlock (hang).
+
+These constraints are mitigated by:
+- the timeouts on network messages, block downloads and block verification, which restart verification if it hangs
+- the `Buffer`/`Batch` reservation release when response future is returned by the buffer/batch, even if the future doesn't complete
+  - in general, we should move as much work into futures as possible, unless the design requires sequential `call`s
+- larger `Buffer`/`Batch` bounds
+
+### Buffered Services
+[buffered-services]: #buffered-services
+
+A service should be provided wrapped in a `Buffer` if:
+* it is a complex service
+* it has multiple callers, or
+* it has a single caller than calls it multiple times concurrently.
+
+Services might also have other reasons for using a `Buffer`. These reasons should be documented.
+
+#### Choosing Buffer Bounds
+[choosing-buffer-bounds]: #choosing-buffer-bounds
+
+Zebra's `Buffer` bounds should be set to the maximum number of concurrent requests, plus 1:
+> it's advisable to set bound to be at least the maximum number of concurrent requests the `Buffer` will see
+https://docs.rs/tower/0.4.3/tower/buffer/struct.Buffer.html#method.new
+
+The extra slot protects us from future changes that add an extra caller, or extra concurrency.
+
+As a general rule, Zebra `Buffer`s should all have at least 3 slots (2 + 1), because most Zebra services can
+be called concurrently by:
+* the sync service, and
+* the inbound service.
+
+Services might also have other reasons for a larger bound. These reasons should be documented.
```

Comment:
This might increase in the future with worker tasks for zebra-client blockchain scanning

Reply:
What do you think of this:
```suggestion
As a general rule, Zebra `Buffer`s should all have at least 10 slots, because most Zebra services can
be called concurrently by:
* the sync service,
* the inbound service, and
* multiple concurrent `zebra-client` blockchain scanning tasks.

Services might also have other reasons for a larger bound. These reasons should be documented.
```

Reply:
Actually let's go with 5 slots to limit RAM usage for malicious blocks in the pipeline.